kernel/kernel-api/src/main/java/io/delta/kernel/internal/icebergcompat/IcebergCompatV2MetadataValidatorAndUpdater.java

```diff
@@ -131,9 +132,15 @@ public static void validateDataFileStatus(DataFileStatus dataFileStatus) {
               .newMetadata
               .getPartitionColNames()
               .forEach(
-                  partitonCol -> {
+                  partitionCol -> {
+                    int partitionFieldIndex =
+                        inputContext.newMetadata.getSchema().indexOf(partitionCol);
+                    checkArgument(
+                        partitionFieldIndex != -1,
+                        "Partition column %s not found in the schema",
+                        partitionCol);
                     DataType dataType =
```

Comment:
This is needed to avoid a NPE and surface a better failure message in case the updated schema does not have a partition column. We run this validator/updator before we get to the subsequent validation which also validates against preventing dropped column.

Reply:
NPE? That's scary. What causes that? Do we have a bug in our code?

Reply:
Yeah when IcebergCompatV2 is enabled and this validation is run against a schema which references a partition column which is not defined in the schema, the logic fails with an NPE here. The previous test which tests this doesn't enable icebergCompatV2 so this validation rule never runs and exercises the bug. 

But my current tests do enable icebergCompatV2 and hit this path

Reply:
Ah, yes. The issue here is that our StructType::get is just plain wrong. it allows null pointer exceptions.

@allisonport-db -- we commented this on the clustering PR, too? do you recall what the next step there was? For now I'll add it to our backlog.

Reply:
>Ah, yes. The issue here is that our StructType::get is just plain wrong. it allows null pointer exceptions.

Yeah, I wanted to fix it there to just return null in case the field doesn't exist but it would affect a lot of places.

Reply:
I think it was that StructType::get should throw a good error (not NPE) if the field doesn't exist; I thought we'd maybe do it in that PR but must have missed tracking it